crates/typst-kit/src/package.rs

```diff
@@ -102,14 +98,39 @@ impl PackageStorage {
 
             // Download from network if it doesn't exist yet.
             if spec.namespace == DEFAULT_NAMESPACE {
-                self.download_package(spec, cache_dir, progress)?;
-                if dir.exists() {
-                    return Ok(dir);
-                }
+                return self.download_package(spec, cache_dir, progress);
             }
         }
 
-        Err(PackageError::NotFound(spec.clone()))
+        // In all paths from now, we either succeed or fail with a NotFound.
+        let not_found = |msg| Err(PackageError::NotFound(spec.clone(), msg));
+
+        let Some(packages_dir) = &self.package_path else {
+            // QUESTION: when is this code path hit?
+            return not_found(None);
+        };
+        // We try the fast path
+        let dir = packages_dir.join(&subdir);
+        if dir.exists() {
+            return Ok(dir);
+        }
+        // Fast path failed. Instead we try to diagnose the issue by going
+        // down the subdirectories of the expected path one at a time.
+        let namespace_dir = packages_dir.join(format!("{}", spec.namespace));
+        if !namespace_dir.exists() {
+            return not_found(Some(eco_format!("did not find namespace '{}' while searching at {} for", spec.namespace, namespace_dir.display())));
+        }
+        let package_dir = namespace_dir.join(format!("{}", spec.name));
+        if !package_dir.exists() {
+            return not_found(Some(eco_format!("did not find package '{}' while searching at {} for", spec.name, package_dir.display())));
+        }
+        let version_dir = package_dir.join(format!("{}", spec.version));
+        if !version_dir.exists() {
+            return not_found(Some(eco_format!("did not find version {} while searching at {} for", spec.version, version_dir.display())));
+        }
+        // We'll never hit this path as long as subdir is indeed namespace/name/version.
+        // What's Typst's policy on `unreachable`?
+        unreachable!()
```

Comment:
If it 100% can't ever happen, unreachable is acceptable, but writing the logic in a way that doesn't require it would be much preferred, if possible